Libplanet/Store/TrieStateStore.cs

```diff
@@ -0,0 +1,99 @@
+    /// </summary>
+    public class TrieStateStore : IStateStore
+    {
+        private readonly IKeyValueStore _stateKeyValueStore;
+        private readonly IKeyValueStore _stateHashKeyValueStore;
+
+        /// <summary>
+        /// Creates a new <see cref="TrieStateStore"/>.
+        /// </summary>
+        /// <param name="stateKeyValueStore">The storage to store states. It used by
+        /// <see cref="Trie"/> in internal.</param>
+        /// <param name="stateHashKeyValueStore">The storage to store state hash corresponding to
+        /// block hash.</param>
+        public TrieStateStore(
+            IKeyValueStore stateKeyValueStore, IKeyValueStore stateHashKeyValueStore)
+        {
+            _stateKeyValueStore = stateKeyValueStore;
+            _stateHashKeyValueStore = stateHashKeyValueStore;
+        }
+
+        /// <inheritdoc/>
+        public void SetStates<T>(
+            HashDigest<SHA256> blockHash,
+            IImmutableDictionary<string, IValue> states,
+            Func<HashDigest<SHA256>, Block<T>> blockGetter)
+            where T : IAction, new()
+        {
+            var block = blockGetter(blockHash);
+            var previousBlockStateHashBytes = block?.PreviousHash is null
+                ? null
+                : _stateHashKeyValueStore.Get(block.PreviousHash.Value.ToByteArray());
+            var trieRoot = previousBlockStateHashBytes is null
+                ? null
+                : new HashNode(new HashDigest<SHA256>(previousBlockStateHashBytes));
+
+            var prevStatesTrie = new Trie.Trie(_stateKeyValueStore, trieRoot);
+
+            foreach (var pair in states)
+            {
+                prevStatesTrie.Set(Encoding.UTF8.GetBytes(pair.Key), pair.Value);
+            }
+
+            var newStateTrie = prevStatesTrie.Commit();
+            _stateHashKeyValueStore.Set(
+                blockHash.ToByteArray(), newStateTrie.Root.Hash().ToByteArray());
+        }
+
+        /// <inheritdoc/>
+        public IValue GetState(
+            string stateKey,
+            HashDigest<SHA256>? blockHash = null,
+            Guid? chainId = null)
+        {
+            var stateHash = _stateHashKeyValueStore.Get(blockHash?.ToByteArray());
+            var stateTrie = new Trie.Trie(
+                _stateKeyValueStore, new HashNode(new HashDigest<SHA256>(stateHash)));
+            var key = Encoding.UTF8.GetBytes(stateKey);
+            return stateTrie.TryGet(key, out IValue value) ? value : null;
+        }
+
+        /// <inheritdoc/>
+        public bool ContainsBlockStates(HashDigest<SHA256> blockHash)
+        {
+            return _stateHashKeyValueStore.Exists(blockHash.ToByteArray());
+        }
+
+        /// <inheritdoc/>
+        public void ForkStates<T>(Guid sourceChainId, Guid destinationChainId, Block<T> branchpoint)
+            where T : IAction, new()
+        {
+            // Do nothing.
```

Comment:
Is it okay that we don't anything here even throwing / logging error?

Reply:
Actually, `ForkStates(…)` is called in `BlockChain<T>.Fork(…)` so if it throws exceptions it will always fail. And it is intended.